src/cargo/ops/registry/mod.rs

```diff
@@ -116,12 +119,13 @@ impl RegistryCredentialConfig {
 /// * `token_required`: If `true`, the token will be set.
 fn registry(
     gctx: &GlobalContext,
+    package_id: Option<PackageId>,
     token_from_cmdline: Option<Secret<&str>>,
     reg_or_index: Option<&RegistryOrIndex>,
     force_update: bool,
     token_required: Option<Operation<'_>>,
-) -> CargoResult<(Registry, RegistrySourceIds)> {
-    let source_ids = get_source_id(gctx, reg_or_index)?;
+) -> CargoResult<(Registry, bool, RegistrySourceIds)> {
+    let (use_package_source_id, source_ids) = get_source_id(gctx, package_id, reg_or_index)?;
```

Comment:
Not thrilled with this change.

As I mentioned in the other comment, it seems like we use some parameters only to call `get_source_id`.  What if instead we moved `get_source_id` out of this function, passing whats needed into it?

Reply:
Sounds good!